### PR TITLE
validate codes list for saved codes tab

### PR DESCRIPTION
**Pull request by sb-prateekkeerthi**
## Terminology Service PR

Jira Ticket: [MAT-6856](https://jira.cms.gov/browse/MAT-6856)
(Optional) Related Tickets:

### Summary
validate codes list for saved codes tab

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java [hunk 1]
```diff
@@ -85,4 +86,13 @@ public ResponseEntity<Code> getCode(
     return ResponseEntity.ok()
         .body(fhirTerminologyService.retrieveCode(code, codeSystem, version, user.getApiKey()));
   }
+
+  @PostMapping(path = "/codesList", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java]: Preferred: `/codes` to indicate it is a collection of `Code` resources.

File: src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java [hunk 2]
```diff
@@ -85,4 +86,13 @@ public ResponseEntity<Code> getCode(
...
+  public ResponseEntity<List<Code>> getCodesList(
+      @RequestBody() List<Map<String, String>> codeList, Principal principal) {
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java:92]: Suggestion: Rename `codeList` to better explain what the Strings are in the Map. For instance, if they are the code and code system, then `codeAndCodeSystems` perhaps. 

I got slightly confused by the method named `getCodesList` taking an argument called codeList, since that sounds like the expected return object.

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 3]
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+              String codeName = code.get("code");
+              String codeSystemName = code.get("codeSystem");
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  oid != null
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
+
+              if (mappedVersion != null) {
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]:
```suggestion
             if (mappedVersion.isPresent()) {
```
Recommend `isPresent` since `mappedVersion` is an `Optional`

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 4]
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+            code -> {
+              List<CodeSystemEntry> codeSystemEntries = mappingService.getCodeSystemEntries();
+              String codeName = code.get("code");
+              String codeSystemName = code.get("codeSystem");
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  oid != null
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: Suggest moving the null check to the top of `mapToFhirVersion()`.

Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]:
```suggestion
                  oid != null
                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
                      : Optional.empty();
```
Alternatively to moving the null check inside `mapToFhirVersion`, keep with the Optional pattern and return `Optional.empty()`.

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 5]
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+              .map(
+                  codeSystemVersion ->
+                      Map.entry(codeSystemVersion.getVsac(), codeSystemVersion.getFhir()))
+              .findFirst();
+    }
+
+    if (result.isPresent()) {
+      return result;
+    }
+    return null;
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:298]:
```suggestion
    return Optional.empty();
```
Stick with Optional pattern.

Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]:
```suggestion
    return result;
```
Since `result` is an Optional, it can be returned directly and handled by the `mappedVersion.isPresent()` check.

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 6]
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+              }
+
+              return retrieveCodes(codeName, codeSystemName, null, null, oid, apiKey);
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Optional<Map.Entry<String, String>> mapToFhirVersion(
+      String version, String oid, List<CodeSystemEntry> codeSystemEntries) {
+
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]:
```suggestion
    if (oid == null) {
      return Optional.empty();
    }
```
Recommend moving the null check to here to avoid having to check for null before every usage of this method (which, so far, there's only 1).

File: src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java (hunk 1 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java]: thanks for the comment I have changed it to /codes

File: src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java (hunk 2 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java:92]: thanks for the comment I have changed it to getCodesAndCodeSystems

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java (hunk 3 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: thank you. I have made the changes

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java (hunk 5 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: thank you. I have made the changes

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java (hunk 6 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: I moved it to here to make it more generic

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java (hunk 4 above)
Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: thank you, I have made those changes

Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: thank you, I have made those changes

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 7]
```diff
@@ -278,4 +277,104 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+              .map(
+                  codeSystemVersion ->
+                      Map.entry(codeSystemVersion.getVsac(), codeSystemVersion.getFhir()))
+              .findFirst();
+    }
+
+    if (result.isPresent()) {
+      return result;
+    }
+    return result;
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]:
```suggestion
    return result;
```
Can be simplified to the above.

Comment (sb-prateekkeerthi) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: thank you, I just fixed it

File: src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java [hunk 8]
```diff
@@ -278,4 +277,104 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
...
+      }
+      String codeJson =
+          fhirTerminologyServiceWebClient.getCodeResource(codeName, codeSystem, apiKey);
+
+      Parameters parameters = fhirContext.newJsonParser().parseResource(Parameters.class, codeJson);
+      Code code =
+          Code.builder()
+              .name(codeName)
+              .codeSystem(codeSystemName)
+              .version(vsacVersion)
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: Should pass `fhirVersion`. `vsacService.getCodeStatus` compares `version` against the fhirVersion in the mapping doc to look up the SVS version (which is known at this point)
```suggestion
              .version(fhirVersion)
```

Comment (jkotanchik-SB) [reply src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java]: Turns out, we needed both versions, so I added an `svsVersion` entry to the DTO.

**Review by jkotanchik-SB — APPROVED** [review]
